Difficulty running tests locally

I'm trying to reproduce the travis.ci script locally with htcluster and stumbling a bit

```
mrocklin@carbon:~/workspace/hadoop-test-cluster$ htcluster startup --image base:latest --mount ~/workspace/skein/:skein
Starting cluster with image jcrist/hadoop-testing-base:latest ...
Creating network "example.com" with the default driver
Creating edge   ... done
Creating worker ... done
Creating master ... done
mrocklin@carbon:~/workspace/hadoop-test-cluster$ cd ../skein/
mrocklin@carbon:~/workspace$ htcluster login
p[testuser@edge ~]$ ls
miniconda  skein
[testuser@edge ~]$ ./skein/continuous_integration/travis/docker_install.sh 3.6
+ packages='grpcio pyyaml cryptography pytest flake8'
+ [[ 3.6 == \2\.\7 ]]
+ conda install grpcio pyyaml cryptography pytest flake8
Solving environment: done

## Package Plan ##

  environment location: /home/testuser/miniconda

  added / updated specs: 
    - cryptography
    - flake8
    - grpcio
    - pytest
    - pyyaml


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    pyyaml-3.12                |   py36hafb9ca4_1         159 KB
    py-1.5.3                   |           py36_0         135 KB
    flake8-3.5.0               |           py36_1         122 KB
    cryptography-2.2.2         |   py36h14c3975_0         599 KB
    openssl-1.0.2o             |       h20670df_0         3.4 MB
    pyflakes-1.6.0             |   py36h7bd6a15_0          85 KB
    conda-4.5.4                |           py36_0         1.0 MB
    attrs-18.1.0               |           py36_0          43 KB
    pytest-3.6.2               |           py36_0         300 KB
    pluggy-0.6.0               |   py36hb689045_0          23 KB
    grpcio-1.12.1              |   py36hdbcaa40_0         1.7 MB
    certifi-2018.4.16          |           py36_0         142 KB
    mccabe-0.6.1               |   py36h5ad9710_1          13 KB
    more-itertools-4.2.0       |           py36_0          78 KB
    pycodestyle-2.3.1          |   py36hf609f19_0          55 KB
    atomicwrites-1.1.5         |           py36_0          10 KB
    ca-certificates-2018.03.07 |                0         124 KB
    ------------------------------------------------------------
                                           Total:         8.0 MB

The following NEW packages will be INSTALLED:

    atomicwrites:    1.1.5-py36_0         
    attrs:           18.1.0-py36_0        
    flake8:          3.5.0-py36_1         
    grpcio:          1.12.1-py36hdbcaa40_0
    mccabe:          0.6.1-py36h5ad9710_1 
    more-itertools:  4.2.0-py36_0         
    pluggy:          0.6.0-py36hb689045_0 
    py:              1.5.3-py36_0         
    pycodestyle:     2.3.1-py36hf609f19_0 
    pyflakes:        1.6.0-py36h7bd6a15_0 
    pytest:          3.6.2-py36_0         
    pyyaml:          3.12-py36hafb9ca4_1  

The following packages will be UPDATED:

    ca-certificates: 2017.08.26-h1d4fec5_0 --> 2018.03.07-0        
    certifi:         2018.1.18-py36_0      --> 2018.4.16-py36_0    
    conda:           4.4.11-py36_0         --> 4.5.4-py36_0        
    cryptography:    2.1.4-py36hd09be54_0  --> 2.2.2-py36h14c3975_0
    openssl:         1.0.2n-hb7f436b_0     --> 1.0.2o-h20670df_0   

Proceed ([y]/n)? y


Downloading and Extracting Packages
pyyaml 3.12: 100% 1.0/1 [00:00<00:00,  6.65it/s] 
py 1.5.3: 100% 1.0/1 [00:00<00:00,  7.92it/s]               
flake8 3.5.0: 100% 1.0/1 [00:00<00:00,  6.48it/s]               
cryptography 2.2.2: 100% 1.0/1 [00:00<00:00,  2.30it/s]               
openssl 1.0.2o: 100% 1.0/1 [00:02<00:00,  2.11s/it]               
pyflakes 1.6.0: 100% 1.0/1 [00:00<00:00, 12.22it/s]
conda 4.5.4: 100% 1.0/1 [00:00<00:00,  1.31it/s]               
attrs 18.1.0: 100% 1.0/1 [00:00<00:00,  4.55it/s]                
pytest 3.6.2: 100% 1.0/1 [00:00<00:00,  4.40it/s] 
pluggy 0.6.0: 100% 1.0/1 [00:00<00:00,  9.45it/s]   
grpcio 1.12.1: 100% 1.0/1 [00:01<00:00,  1.43s/it]                
certifi 2018.4.16: 100% 1.0/1 [00:00<00:00, 12.44it/s]
mccabe 0.6.1: 100% 1.0/1 [00:00<00:00, 26.57it/s]
more-itertools 4.2.0: 100% 1.0/1 [00:00<00:00, 12.23it/s]
pycodestyle 2.3.1: 100% 1.0/1 [00:00<00:00, 19.62it/s]
atomicwrites 1.1.5: 100% 1.0/1 [00:00<00:00, 28.57it/s]
ca-certificates 2018.03.07: 100% 1.0/1 [00:00<00:00, 16.11it/s]
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
+ pip install grpcio-tools
Collecting grpcio-tools
  Downloading https://files.pythonhosted.org/packages/58/28/ce680c789fddd85ccf7fd1b9c11a26b9802b32c9cfaccf7d2cc0ee3bc276/grpcio_tools-1.13.0-cp36-cp36m-manylinux1_x86_64.whl (22.2MB)
    100% |████████████████████████████████| 22.2MB 67kB/s 
Collecting grpcio>=1.13.0 (from grpcio-tools)
  Downloading https://files.pythonhosted.org/packages/36/13/b05bbd05d9f0c19e5746f149a285aa81f7c353e231ffecdb237c6e2ad3cc/grpcio-1.13.0-cp36-cp36m-manylinux1_x86_64.whl (9.1MB)
    100% |████████████████████████████████| 9.1MB 154kB/s 
Collecting protobuf>=3.5.0.post1 (from grpcio-tools)
  Downloading https://files.pythonhosted.org/packages/fc/f0/db040681187496d10ac50ad167a8fd5f953d115b16a7085e19193a6abfd2/protobuf-3.6.0-cp36-cp36m-manylinux1_x86_64.whl (7.1MB)
    100% |████████████████████████████████| 7.1MB 202kB/s 
Requirement already satisfied: six>=1.5.2 in ./miniconda/lib/python3.6/site-packages (from grpcio>=1.13.0->grpcio-tools)
Requirement already satisfied: setuptools in ./miniconda/lib/python3.6/site-packages (from protobuf>=3.5.0.post1->grpcio-tools)
Installing collected packages: grpcio, protobuf, grpcio-tools
  Found existing installation: grpcio 1.12.1
    Uninstalling grpcio-1.12.1:
      Successfully uninstalled grpcio-1.12.1
Successfully installed grpcio-1.13.0 grpcio-tools-1.13.0 protobuf-3.6.0
You are using pip version 9.0.1, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
+ cd /home/testuser/skein
+ pip install -v --no-deps -e .
Obtaining file:///home/testuser/skein
  Running setup.py (path:/home/testuser/skein/setup.py) egg_info for package from file:///home/testuser/skein
    Running command python setup.py egg_info
    running egg_info
    writing skein.egg-info/PKG-INFO
    writing dependency_links to skein.egg-info/dependency_links.txt
    writing entry points to skein.egg-info/entry_points.txt
    writing requirements to skein.egg-info/requires.txt
    writing top-level names to skein.egg-info/top_level.txt
    reading manifest file 'skein.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    writing manifest file 'skein.egg-info/SOURCES.txt'
  Source in /home/testuser/skein has version 0.0.1+152.g7d46c28, which satisfies requirement skein==0.0.1+152.g7d46c28 from file:///home/testuser/skein
Installing collected packages: skein
  Running setup.py develop for skein
    Running command /home/testuser/miniconda/bin/python -c "import setuptools, tokenize;__file__='/home/testuser/skein/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" develop --no-deps
    running develop
    running egg_info
    writing skein.egg-info/PKG-INFO
    writing dependency_links to skein.egg-info/dependency_links.txt
    writing entry points to skein.egg-info/entry_points.txt
    writing requirements to skein.egg-info/requires.txt
    writing top-level names to skein.egg-info/top_level.txt
    reading manifest file 'skein.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    writing manifest file 'skein.egg-info/SOURCES.txt'
    running build_ext
    Creating /home/testuser/miniconda/lib/python3.6/site-packages/skein.egg-link (link to .)
    Adding skein 0.0.1+152.g7d46c28 to easy-install.pth file
    Installing skein script to /home/testuser/miniconda/bin

    Installed /home/testuser/skein
Successfully installed skein
Cleaning up...
You are using pip version 9.0.1, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
+ conda list
# packages in environment at /home/testuser/miniconda:
#
# Name                    Version                   Build  Channel
asn1crypto                0.24.0                   py36_0  
atomicwrites              1.1.5                    py36_0  
attrs                     18.1.0                   py36_0  
ca-certificates           2018.03.07                    0  
certifi                   2018.4.16                py36_0  
cffi                      1.11.4           py36h9745a5d_0  
chardet                   3.0.4            py36h0f667ec_1  
conda                     4.5.4                    py36_0  
conda-env                 2.6.0                h36134e3_1  
cryptography              2.2.2            py36h14c3975_0  
flake8                    3.5.0                    py36_1  
grpcio                    1.12.1           py36hdbcaa40_0  
grpcio                    1.13.0                    <pip>
grpcio-tools              1.13.0                    <pip>
idna                      2.6              py36h82fb2a8_1  
libedit                   3.1                  heed3624_0  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 7.2.0                h7cc24e2_2  
libstdcxx-ng              7.2.0                h7a57d05_2  
mccabe                    0.6.1            py36h5ad9710_1  
more-itertools            4.2.0                    py36_0  
ncurses                   6.0                  h9df7e31_2  
openssl                   1.0.2o               h20670df_0  
pip                       9.0.1            py36h6c6f9ce_4  
pluggy                    0.6.0            py36hb689045_0  
protobuf                  3.6.0                     <pip>
py                        1.5.3                    py36_0  
pycodestyle               2.3.1            py36hf609f19_0  
pycosat                   0.6.3            py36h0a5515d_0  
pycparser                 2.18             py36hf9f622e_1  
pyflakes                  1.6.0            py36h7bd6a15_0  
pyopenssl                 17.5.0           py36h20ba746_0  
pysocks                   1.6.7            py36hd97a5b1_1  
pytest                    3.6.2                    py36_0  
python                    3.6.4                hc3d631a_1  
pyyaml                    3.12             py36hafb9ca4_1  
readline                  7.0                  ha6073c6_4  
requests                  2.18.4           py36he2e5f8d_1  
ruamel_yaml               0.15.35          py36h14c3975_1  
setuptools                38.4.0                   py36_0  
six                       1.11.0           py36h372c433_1  
skein                     0.0.1+152.g7d46c28           <pip>
sqlite                    3.22.0               h1bed415_0  
tk                        8.6.7                hc745277_3  
urllib3                   1.22             py36hbe7ace6_0  
wheel                     0.30.0           py36hfd4bba0_1  
xz                        5.2.3                h55aa19d_2  
yaml                      0.1.7                had09818_2  
zlib                      1.2.11               ha838bed_2  


[testuser@edge ~]$ export CONDA_ENV="/home/testuser/miniconda/"
[testuser@edge ~]$ $CONDA_ENV/bin/py.test skein/ -vv
============================= test session starts ==============================
platform linux -- Python 3.6.4, pytest-3.6.2, py-1.5.3, pluggy-0.6.0 -- /home/testuser/miniconda/bin/python
cachedir: skein/.pytest_cache
rootdir: /home/testuser/skein, inifile:
collected 0 items / 1 errors                                                   

==================================== ERRORS ====================================
______________________________ ERROR collecting  _______________________________
miniconda/lib/python3.6/site-packages/_pytest/config/__init__.py:369: in _getconftestmodules
    return self._path2confmods[path]
E   KeyError: local('/home/testuser/skein/skein/test')

During handling of the above exception, another exception occurred:
miniconda/lib/python3.6/site-packages/_pytest/config/__init__.py:400: in _importconftest
    return self._conftestpath2mod[conftestpath]
E   KeyError: local('/home/testuser/skein/skein/test/conftest.py')

During handling of the above exception, another exception occurred:
miniconda/lib/python3.6/site-packages/_pytest/config/__init__.py:406: in _importconftest
    mod = conftestpath.pyimport()
miniconda/lib/python3.6/site-packages/py/_path/local.py:668: in pyimport
    __import__(modname)
skein/skein/__init__.py:1: in <module>
    from .core import Client, Application, ApplicationClient, Security
skein/skein/core.py:17: in <module>
    from . import proto
skein/skein/proto/__init__.py:3: in <module>
    from .skein_pb2 import (Empty, FinalStatus, ApplicationState, Resources, File,
E   ImportError: cannot import name 'ApplicationSpec'

During handling of the above exception, another exception occurred:
miniconda/lib/python3.6/site-packages/py/_path/common.py:377: in visit
    for x in Visitor(fil, rec, ignore, bf, sort).gen(self):
miniconda/lib/python3.6/site-packages/py/_path/common.py:429: in gen
    for p in self.gen(subdir):
miniconda/lib/python3.6/site-packages/py/_path/common.py:418: in gen
    dirs = self.optsort([p for p in entries
miniconda/lib/python3.6/site-packages/py/_path/common.py:419: in <listcomp>
    if p.check(dir=1) and (rec is None or rec(p))])
miniconda/lib/python3.6/site-packages/_pytest/main.py:511: in _recurse
    ihook = self.gethookproxy(path)
miniconda/lib/python3.6/site-packages/_pytest/main.py:413: in gethookproxy
    my_conftestmodules = pm._getconftestmodules(fspath)
miniconda/lib/python3.6/site-packages/_pytest/config/__init__.py:383: in _getconftestmodules
    mod = self._importconftest(conftestpath)
miniconda/lib/python3.6/site-packages/_pytest/config/__init__.py:412: in _importconftest
    raise ConftestImportFailure(conftestpath, sys.exc_info())
E   _pytest.config.ConftestImportFailure: ImportError("cannot import name 'ApplicationSpec'",)
E     File "/home/testuser/skein/skein/__init__.py", line 1, in <module>
E       from .core import Client, Application, ApplicationClient, Security
E     File "/home/testuser/skein/skein/core.py", line 17, in <module>
E       from . import proto
E     File "/home/testuser/skein/skein/proto/__init__.py", line 3, in <module>
E       from .skein_pb2 import (Empty, FinalStatus, ApplicationState, Resources, File,
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.33 seconds ============================
```

--- a/docs/source/develop.rst
+++ b/docs/source/develop.rst
@@ -1,0 +1,199 @@
+Development
+===========
+
+This page provides information on how to build, test, and develop ``Skein``.
+
+
+Building Skein
+--------------
+
+Clone Repository
+~~~~~~~~~~~~~~~~
+
+Clone the skein git repository:
+
+.. code-block:: console
+
+    $ git clone https://github.com/jcrist/skein.git
+
+
+Install Dependencies (Conda)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We recommend using the Conda_ package manager to setup your development
+environment. Here we setup a conda environment to contain the build
+dependencies.
+
+.. code-block:: console
+
+    # Create a new conda environment
+    $ conda create -n skein
+
+    # Activate environment
+    $ conda activate skein
+
+    # Install dependencies
+    $ conda install -c conda-forge grpcio protobuf cryptography pyyaml
+
+    # Install grpcio-tools (not on conda-forge currently)
+    $ pip install grpcio-tools
+
+
+Besides the above dependencies, you'll also need Maven_. You can install Maven
+using your system package manager, the maven website, or use Conda:
+
+.. code-block:: console
+
+    $ conda install -c conda-forge maven
+
+
+Install Dependencies (Pip)
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can also setup the development environment using ``pip``.
+
+.. code-block:: console
+
+    $ pip install grpcio protobuf cryptography pyyaml grpcio-tools
+
+
+Besides the above dependencies, you'll also need Maven_. You can install Maven
+using your system package manager or via the maven website.
+
+
+Build and Install Skein
+~~~~~~~~~~~~~~~~~~~~~~~
+
+You can build and install Skein as an editable package or a regular install.
+
+.. code-block:: console
+
+    # Build and install skein as an editable package
+    $ python setup.py develop
+
+    # or, build and install as a regular package
+    $ python setup.py install
+
+
+Running the Tests
+-----------------
+
+The test suite is designed to run in a specific hadoop setup, provided by the
+`hadoop-test-cluster`_ package. This is a CLI tool for setting up a Hadoop
+cluster using `docker compose`_. This requires ``docker compose`` be installed,
+and the docker daemon already be running. Please follow the install
+instructions for your system `here
+<https://docs.docker.com/compose/install/>`__.
+
+
+Install `hadoop-test-cluster`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can install ``hadoop-test-cluster`` using ``pip``. This assumes you already
+have ``docker`` and ``docker-compose`` already installed.
+
+The `hadoop-test-cluster`_ repository readme has documentation on usage - below
+we provide a few commands needed for using the cluster to run the tests.
+
+.. code-block:: console
+
+    $ pip install hadoop-test-cluster
+
+
+Startup the Test Cluster
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+This command starts up a tiny Hadoop cluster with ``simple`` security, and
+mounts the current directory as ``~/skein`` on every node. To create a cluster
+with ``kerberos`` security enabled, use ``--image kerberos`` instead.
+
+.. code-block:: console
+
+    $ htcluster startup --image base --mount .:skein
+
+
+Login to the Edge Node
+~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: console
+
+    $ htcluster login
+
+
+Setup the Development Environment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The docker image already has Conda_ installed. After startup, you only need to
+install the runtime and test dependencies (see `Install Dependencies (Conda)`_).
+Alternatively, Maven_ is also already installed on the docker image, so you can
+skip the instructions for building Skein locally above and do everything on the
+docker image.
+
+You also need ``pytest`` to run the tests, and ``flake8`` to run the lint
+checks.
+
+.. code-block:: console
+
+    $ conda install -c conda-forge pytest flake8
+
+
+Build and Install Skein
+~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: console
+
+    $ python setup.py develop
+
+
+Run the Tests
+~~~~~~~~~~~~~
+
+.. code-block:: console
+
+    $ pytest skein
+
+
+Run the Linter
+~~~~~~~~~~~~~~
+
+.. code-block:: console
+
+    $ flake8 skein
+
+
+Shutdown the Cluster
+~~~~~~~~~~~~~~~~~~~~
+
+When you are done developing, you can shutdown the cluster using the following
+
+.. code-block:: console
+
+    $ htcluster shutdown
+
+
+Building the Documentation
+--------------------------
+
+Skein uses Sphinx_ for documentation. The source files are located in
+``skein/docs/source``. To build the documentation locally, first install the
+documentation build requirements
+
+.. code-block:: console
+
+    $ pip install sphinx numpydoc sphinxcontrib.autoprogram
+
+Then build the documentation with ``make``
+
+.. code-block:: console
+
+    # Running from the skein/docs folder
+    $ make html
+
+The resulting HTML files end up in the ``build/html`` directory.
+
+
+.. _Conda: https://conda.io/docs/
+.. _Maven: http://maven.apache.org/
+.. _Sphinx: http://www.sphinx-doc.org/
+.. _docker compose: https://docs.docker.com/compose/
+.. _hadoop-test-cluster: https://github.com/jcrist/hadoop-test-cluster

--- a/docs/source/index.rst
+++ b/docs/source/index.rst
@@ -62,3 +62,11 @@ User Documentation
     recipes.rst
     api.rst
     cli.rst
+
+Developer Documentation
+-----------------------
+
+.. toctree::
+    :maxdepth: 2
+
+    develop.rst

--- a/docs/source/recipes-ipython-kernel.rst
+++ b/docs/source/recipes-ipython-kernel.rst
@@ -1,7 +1,7 @@
 Remote IPython Kernel
 =====================
 
-The :mod:`skein.recipes.ipython_kernel` module provides a commandline recipe
+The :mod:`skein.recipes.ipython_kernel` module provides a command-line recipe
 for starting a remote IPython kernel on a YARN container. The intended use is
 to execute the module in a service, using the command:
 

--- a/docs/source/recipes-jupyter-notebook.rst
+++ b/docs/source/recipes-jupyter-notebook.rst
@@ -1,7 +1,7 @@
 Jupyter Notebook Server
 =======================
 
-The :mod:`skein.recipes.jupyter_notebook` module provides a commandline recipe
+The :mod:`skein.recipes.jupyter_notebook` module provides a command-line recipe
 for starting a Jupyter notebook server on a YARN container. The intended use is
 to execute the module in a service, using the command:
 
@@ -85,8 +85,8 @@ computation.
     [########################################] | 100% Completed | 16.1s
 
 
-Fix the Packagd Kernelspec
-^^^^^^^^^^^^^^^^^^^^^^^^^^
+Fix the Packaged Kernelspec
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 When ``conda-pack`` packages an environment, it trys to fix as many absolute
 paths found in the environment to be as relocatable as possible. Those that

